src/wallet/wallet_rpc_server.cpp

```diff
@@ -3089,6 +3089,171 @@ namespace tools
+    
+    // check the given seed
+    {
+      if (!crypto::ElectrumWords::words_to_bytes(req.seed, m_recovery_key, old_language))
+      {
+        er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+        er.message = "Electrum-style word list failed verification";
+        return false;
+      }
+    }
+
+    // process seed_offset if given
+    {
+      if (!req.seed_offset.empty())
+      {
+        m_recovery_key = cryptonote::decrypt_key(m_recovery_key, req.seed_offset);
+      }
+    }
+    {
+      po::options_description desc("dummy");
+      const command_line::arg_descriptor<std::string, true> arg_password = {"password", "password"};
+      const char *argv[4];
+      int argc = 3;
+      argv[0] = "wallet-rpc";
+      argv[1] = "--password";
+      argv[2] = req.password.c_str();
+      argv[3] = NULL;
+      vm2 = *m_vm;
+      command_line::add_arg(desc, arg_password);
+      po::store(po::parse_command_line(argc, argv, desc), vm2);
+    }
+
+    auto rc = tools::wallet2::make_new(vm2, true, nullptr);
+    std::unique_ptr<wallet2> wal;
+    wal = std::move(rc.first);
+    if (!wal)
+    {
+      er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+      er.message = "Failed to create wallet";
+      return false;
+    }
+
+    epee::wipeable_string password = rc.second.password();
+   
+
+    bool was_deprecated_wallet = ((old_language == crypto::ElectrumWords::old_language_name) ||
+    crypto::ElectrumWords::get_is_old_style_seed(req.seed));
+
+    std::string mnemonic_language = old_language;
+    std::string m_mnemonic_language;
+    std::vector<std::string> language_list;
+    crypto::ElectrumWords::get_language_list(language_list);
+    if (mnemonic_language.empty() && std::find(language_list.begin(), language_list.end(), m_mnemonic_language) != language_list.end())
+    {
+      mnemonic_language = m_mnemonic_language;
+    }
+
+    if (((mnemonic_language.empty() || mnemonic_language == crypto::ElectrumWords::old_language_name) && was_deprecated_wallet))
+    {
+      if (was_deprecated_wallet)
+      {
+        // The user had used an older version of the wallet with old style mnemonics.
+        LOG_PRINT_L0(tools::wallet_rpc_server::tr("You had been using "
+          "a deprecated version of the wallet. Please use the new seed that we provide.\n"));
+      }
+      if (req.language.empty())
+      {
+        er.code = WALLET_RPC_ERROR_CODE_UNKNOWN_ERROR;
+        er.message = "Wallet is deprecated. Would provide new seed. Please specify language to use for new seed.";
+        return false;
+      }
+      mnemonic_language = req.language;
+      if (mnemonic_language.empty())
+        return {};
+    }
+
+    wal->set_seed_language(mnemonic_language);
+
+    crypto::secret_key recovery_val;
+    try
+    {
+      recovery_val = wal->generate(m_wallet_file, std::move(rc.second).password(), m_recovery_key, true, false, false);
+      LOG_PRINT_L0(tools::wallet_rpc_server::tr("Wallet has been restored.\n"));
```

Comment:
Please use MERROR/MWARNING/MINFO/MDEBUG in new code. In this case, MINFO seems best.
No need to use tr() here, it's for translation, and only simplewallet is translated.